introducing-pgzx.mdx

```diff
@@ -0,0 +1,208 @@
+
+To trigger the bug, you would call something like this:
+
+```sql
+postgres=# select char_count_zig('hi hii', '');
+
+server closed the connection unexpectedly
+        This probably means the server terminated abnormally
+        before or while processing the request.
+The connection to the server was lost. Attempting reset: Failed.
+```
+
+If this was C code, this would be segmentation fault or undefined behavior. If you try this with the `char_count_zig` extension, the Postgres process still crashes (not the full server, just the process serving the connection), but if you check the logs you'll see an error message like this:
+
+```
+thread 70501513 panic: index out of bounds: index 0, len 0
+/Users/tsg/src/xataio/pgzx/examples/char_count_zig/src/main.zig:21:32: 0x103aaedff in char_count_zig (char_count_zig)
+        if (char == target_char[0]) {
+                               ^
+/Users/tsg/src/xataio/pgzx/src/pgzx/fmgr.zig:95:5: 0x103aaf20f in call (char_count_zig)
+    const value = @call(.no_async, impl, callArgs) catch |e| elog.throwAsPostgresError(src, e);
+    ^
+???:?:?: 0x10316045b in _ExecInterpExpr (???)
+???:?:?: 0x10315fbef in _ExecInterpExprStillValid (???)
+???:?:?: 0x10326ceef in _evaluate_expr (???)
+???:?:?: 0x10326da67 in _simplify_function (???)
+???:?:?: 0x10326bacf in _eval_const_expressions_mutator (???)
+```
+
+It points exactly where the error happened. This happens because Zig has runtime checks depending on the [build mode](https://ziglang.org/documentation/master/#Build-Mode). The `ReleaseSafe` mode, for example, trades a bit of performance for more safety checks.
+
+It's worth pointing out that the stacktrace worked so well because the error was in Zig code. When building Postgres extensions, you often have to call Postgres APIs and you can still cause segmentation faults if you use them incorrectly.
+
+### Memory management
+
+Postgres uses a [memory context system](https://github.com/postgres/postgres/blob/master/src/backend/utils/mmgr/README) to manage memory. Memory allocated in a "context" can be freed all at once (for example, when a query execution is finished), which simplifies memory management significantly, because you only need to track contexts, not individual allocations. Contexts are also hierarchical, so you can create a context that is a child of another context, and when the parent context is freed, all children are freed as well.
+
+Another advantage of memory contexts is that they improve memory monitoring, because context have names and you can see how much memory is used by each context. This is useful for debugging memory leaks, for example.
```

Comment:
See: https://www.postgresql.org/docs/current/view-pg-backend-memory-contexts.html

I never have had a look at that table 😆 
Maybe I should give it a try.

Reply:
Adding the link to it.